_posts/en/2016-12-28-delegates-closures.md

```diff
@@ -0,0 +1,299 @@
+
+**Swift and Objective-C**
+
+For those who don't know, in iOS, you can use two different programming languages : Objective-C and Swift. Everyone is more or less using Swift now, but because of legacy, you can still find some projects made with Objective-C. This is why in this article, every piece of code will be provided as well in Objective-C as in Swift. The two langages have two different syntaxes and in addition to that, even the file structure is different :
+- in Objective-C you have an header (.h file) to declare all the accessible elements, then you will have an implementation file (.m file) that will contain the body of the methods, exactly like in C language.
+- in Swift, you have only one file (.swift file) et the accessibility will be defined depending on "public" or "private".
+
+**Protocol**
+
+Just one last notion, and I promise, we are good to go.
+In this article, I will use a lot the term: "Protocol" : a protocol is an interface in the Apple world, when I say interface, I mean a class that you have to inherits in order to implement some methods. It's a keyword for both Swift and Objective-C.
+
+**Delegates and Closures / Blocks**
+
+In iOS development, you will often find 2 principles : ***delegates***, ***closures (Swift) / blocks (Objective-C)***
+
+We will cover those two points more in detail, but before we go further it is important to know that exist ! Here are the basic principles to know ! Were are now done with the introduction, we can finally go deep and have a proper look !
+
+#### Delegates
+
+A delegate is a reference to an object that we don't know the exact type, but important thing, it inherits a protocol. Because this object inherits a protocol, we then know that we can call the methods defined in the protocol, even if we don"t know in detail the given object. I think that a proper exemple will help us to understand what I am talking about.<br/>
+NB : The code I will provide is just a dummy implementation so that you are able to understand the principles I am talking about, it will not really do any HTTP call on the URL given as a parameter. Let's imagine that I need to do a GET, usually in the mobile development world, we like to handle this with 2 different callbacks for the return of the call. One is for the success case et the other one, is for the error case. Our goal here is to produce a class that will do a GET on a given URL. I want to notify the object that launched this request, if it failed or succeed. In order to avoid a strong dependance, we will use the design pattern of the delegate, thanks to that, I don't need to know the exact type of this object. We will then define a protocol that will contain two methods: onRequestSuccess onRequestFailure. Let's have a look at what it will look like :
```

Comment:
HTTP calls
and the other one
succeeded
we will use the Delegate design pattern